ietf/meeting/views.py

```diff
@@ -2547,10 +2547,16 @@ def session_attendance(request, session_id, num):
         else:
             raise Http404('Bluesheets not found')
 
+    person = request.user.person
+    if request.method=='POST':
+        session.attended_set.get_or_create(person=person, origin="self declared")
```

Comment:
First point is totally legit, but I don't understand the second point. Thinking this through, we might have a race if MeetEcho calls `api_add_session_attendees` with the user pk, after the user GETs the form but before they POST. Doing the `can_add` check before accepting the POST means that it will now fail the check (there's now an existing `Attendance` record).

But I also don't understand `defaults` in this context. Are you suggesting something like what appears in `nomcom.views.volunteer`?

Reply:
> we might have a race if MeetEcho calls api_add_session_attendees with the user pk, after the user GETs the form but before they POST. Doing the can_add check before accepting the POST means that it will now fail the check (there's now an existing Attendance record).

That is the race I'm thinking about. Adding the `can_add` check to the POST significantly narrows the window, but there's still a race because MeetEcho's API call might happen just after testing `can_add`. What I'm suggesting is like the volunteer view you mention, specifically:
```python
session.attended_set.get_or_create(person=person, defaults={"origin": "self declared"})
```
The defaults are used only when creating a record because wasn't found - this is equivalent to trying a `.filter(person=person).get()`, then if it fails, creating a person with `(person=person, origin="self declared")`. It has the advantage that it enforces that check in the DB, so it closes the race completely.

I'm assuming that the other side of this also has a guard against this race. It might not, in which case in this scheme I think it should get a guard.

Alternatively, we could just allow duplicates. That might be easier. We'd need to be careful to de-dupe when counting and to be careful about priority when displaying the origin - I'd assume that having both a "datatracker" and a "self declared" would be treated as "datatracker" for most purposes since that is the more reliable source.

Reply:
I would prefer to avoid duplicates if at all possible.

Reply:
Apparently that works. Done and tested.

Reply:
The `Attended` model has `unique_together = (('person', 'session'),)`, so presumably django would raise an exception if we try to add a duplicate.

Reply:
Ah, that's great.

It'll turn into a 500 response unless the `IntegrityError` is handled , though. IMO better to race-proof the code. The only reason I can think of to allow this to fail on that is if we wanted to catch it and display a "You were already marked 'attended'" error, but that doesn't seem necessary.

Reply:
However, I don't think the code's using the get_or_create(...,defaults=) pattern yet, and it would be much more robust if it did so. What's the reasning for not doing that?

Reply:
> However, I don't think the code's using the get_or_create(...,defaults=) pattern yet, and it would be much more robust if it did so. What's the reasoning for not doing that?

I did that in 1e7995c67c, sorry if it wasn't clear.

Reply:
Thanks! I didn't review the updates closely enough (or notice the big "Outdated" badge on the comment...)